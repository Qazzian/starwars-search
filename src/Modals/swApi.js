export default class swApi {
	constructor(errorHandler) {
		this.rootUrl = 'https://swapi.co/api/';
		this.errorHandler = errorHandler;
	}

	fetchRequest(url, options = {}) {
		return window.fetch(url, options).then((response) => {
			if (response.ok && response.status === 200) {
				return response.json();
			}
			else {
				throw new Error(`Connection Error: ${response.status}`);
			}
		})
	}

	searchNames(search) {
		this.abortRequest();
		const url = `${this.rootUrl}people/?search=${search}`;
		const options = {
			signal: this.createAbortSignal(),
		};

		return this.fetchRequest(url, options).then((responseJson) => {
			if (responseJson && responseJson.results) {
				return responseJson.results;
			}
			else {
				return [];
			}
		}).catch((error) => {
			return this.handleErrorResponse(error);
		});
	}

	abortRequest() {
		if (this.abortController) {
			this.abortController.abort();
			delete this.abortController;
		}
	}

	createAbortSignal() {
		this.abortController = new window.AbortController();
		return this.abortController.signal;
	}

	handleErrorResponse(error) {
		switch (error.code) {
			case 20:
				return [];
			default:
				if (typeof this.errorHandler === 'function') {
					this.errorHandler(error);
					return [];
				}
				break
		}
	}

	fetchPersonDetails(personUrl) {
		return this.fetchRequest(personUrl).then((responseJson) => {
			const characterData = responseJson;

			return characterData;
		});
	}

	fetchItemNames(urls) {
		const urlList = !Array.isArray(urls) ? [urls] : urls;

		Promise.all(urlList.map((url) => {
			return this.fetchItemName(url);
		}));
	}

	fetchItemName(url) {
		return this.fetchRequest(url).then((responseJson) => {
			if (responseJson && responseJson.name) {
				return responseJson.name;
			}
			return '';
		});
	}
}
