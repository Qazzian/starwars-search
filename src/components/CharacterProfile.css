.App-character-profile {
	text-align: left;
	padding: 0 1rem;
}

.App-character-profile h2 {
	color: rgb(255, 227, 0);
	margin-top: 0;
	margin-bottom: 0.7rem;
}

.App-character-profile p {
	margin: 0.5rem 0;
	padding-left: 1rem;
}

.App-character-profile__species-gender span{
	margin-right: 2rem;
	text-transform: capitalize;
}
