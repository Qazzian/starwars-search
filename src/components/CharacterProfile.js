import React from 'react';

import './CharacterProfile.css';

import ObjectName from './characterProfile/ObjectName';
import FilmList from './characterProfile/FilmList';

class CharacterProfile extends React.Component {
	constructor(props) {
		super(props);

		this.swApi = props.swApi;

		this.state = {
			characterProfile: null,
		}
	}

	render() {

		if (!this.state.characterProfile) {
			return null;
		}

		return (
			<article className="App-character-profile">
				<h2>{this.state.characterProfile.name}</h2>
				<div>
					<p className="App-character-profile__species-gender">
						<span>{this.state.characterProfile.gender}</span>
						{' '}
						<span><ObjectName url={this.state.characterProfile.species} swApi={this.swApi}/></span>

					</p>
					<p>Born: {this.state.characterProfile.birth_year}</p>
					<p>Home world: <ObjectName url={this.state.characterProfile.homeworld} swApi={this.swApi}/></p>
					<FilmList filmList={this.state.characterProfile.films} swApi={this.swApi}/>
				</div>
			</article>
		);
	}

	componentDidMount() {

	}

	componentDidUpdate(prev) {
		if (this.props.characterUrl !== prev.characterUrl) {
			this.getCharacterData();
		}
	}

	getCharacterData() {
		this.swApi.fetchPersonDetails(this.props.characterUrl).then((responses) => {
			console.info('fetchPersonDetails response:', responses);
			this.setState({characterProfile: responses});
		})
	}
}

export default CharacterProfile;
