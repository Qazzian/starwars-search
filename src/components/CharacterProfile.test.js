import React from 'react';
import ReactDOM from 'react-dom';

import CharacterProfile from './CharacterProfile';
import swApi from '../Modals/swApi';

const mockApi = {
	mockPerson: null,
	fetchPersonDetails: (url) => {
		console.info('fetch person', url);
		return Promise.resolve(mockApi.mockPerson);
	}
};

describe('CharacterProfile', () => {
	it('should be defined', (done) => {
		expect(CharacterProfile).toBeDefined();
		done();
	});

	it('should render without errors', (done) => {
		const div = document.createElement('div');
		ReactDOM.render(<CharacterProfile/>, div);
		ReactDOM.unmountComponentAtNode(div);
		done();
	});

	it('should render the profile', (done) => {
		const div = document.createElement('div');
		const mockPerson = {
			name: 'Test Name',
			gender: 'Male',
			species: 'Human',
			homeworld: 'Earth',
			films: ['f1', 'f2', 'f3']
		};
		mockApi.mockPerson = mockPerson;
		ReactDOM.render(<CharacterProfile url={'test1'} swApi={ mockApi }/>, div);
		setTimeout(() => {
			expect(div).toMatchSnapshot('CharacterProfile');
			ReactDOM.unmountComponentAtNode(div);
			done();
		}, 10);
	});
});
