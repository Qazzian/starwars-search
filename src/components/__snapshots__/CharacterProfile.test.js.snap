// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterProfile 1`] = `<div />`;
